src/@utils/Wave.ts

```diff
@@ -0,0 +1,174 @@
+    return layers
+  }
+
+  generateSvg(): Element {
+    const svg = document.createElementNS(SvgWaves.xmlns, 'svg')
+    // TODO: figure out if width,height & xmlns can be removed
+    // could save characters
+    // for example see https://stackoverflow.com/questions/18467982/are-svg-parameters-such-as-xmlns-and-version-needed
+    svg.setAttribute('width', this.properties.width.toString())
+    svg.setAttribute('height', this.properties.height.toString())
+    svg.setAttribute('fill', this.properties.fill ? undefined : 'transparent')
+    svg.setAttribute('xmlns', SvgWaves.xmlns)
+
+    for (let i = 0; i < this.layers.length; i++) {
+      const path = this.generatePath(
+        this.layers[i],
+        this.getOpacityForLayer(i + 1)
+      )
+      svg.appendChild(path)
+    }
+
+    return svg
+  }
+
+  generatePath(points: Point[], opacity = 'ff'): Element {
+    const xPoints = points.map((p) => Math.floor(p.x))
+    const yPoints = points.map((p) => Math.floor(p.y))
+
+    // get bezier control points
+    const xControlPoints = computeControlPoints(xPoints)
+    const yControlPoints = computeControlPoints(yPoints)
+
+    // Should the path be closed & filled or stroke only
+    const closed = this.properties.fill
+
+    // For closed paths define bottom corners as start & end point
+    const bottomLeftPoint = new Point(0, this.properties.height)
+    const bottomRightPoint = new Point(
+      this.properties.width,
+      this.properties.height
+    )
+    const startPoint = closed ? bottomLeftPoint : points[0]
+    const endPoint = closed ? bottomRightPoint : points[points.length - 1]
+
+    // start constructing the 'd' attribute for the <path>
+    let path = `M${startPoint.x},${startPoint.y}`
+
+    // if starting in bottom corner, move to start of wave first
+    if (closed) path += `L${xPoints[0]},${yPoints[0]}`
+
+    // add path curves
+    for (let i = 0; i < xPoints.length - 1; i++) {
+      path +=
+        `C${xControlPoints.p1[i]},${yControlPoints.p1[i]} ` +
+        `${xControlPoints.p2[i]},${yControlPoints.p2[i]} ` +
+        `${xPoints[i + 1]},${yPoints[i + 1]}`
+    }
+
+    // if closing in bottom corner move there and close
+    if (closed) path += `L${endPoint.x},${endPoint.y}Z`
+    // else just close
+    else path += 'AZ'
+
+    // create the path element
+    // TODO: save characters by replacing closing tag with '<path ... />'
+    const svgPath = document.createElementNS(SvgWaves.xmlns, 'path')
+    const colorStyle = closed ? 'fill' : 'stroke'
+    svgPath.setAttributeNS(
+      null,
+      colorStyle,
+      `${this.properties.color}${opacity}`
+    )
+    svgPath.setAttributeNS(null, 'd', path)
+
+    return svgPath
+  }
+
+  getOpacityForLayer(layer: number): string {
+    const { layerCount, maxOpacity, opacitySteps } = this.properties
+
+    const minOpacity = maxOpacity - (layerCount - 1) * opacitySteps
+    const opacityDec = minOpacity + layer * opacitySteps
+
+    return opacityDec.toString(16)
```

Comment:
suggestion(blocking): add comment why we need to transform the string like this (I assume because it needs to be HEX)